src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs

```diff
@@ -408,16 +435,28 @@ private PingReply GetAddressAndSend(string hostNameOrAddress, int timeout, byte[
             }
         }
 
-        private async Task<PingReply> GetAddressAndSendAsync(string hostNameOrAddress, int timeout, byte[] buffer, PingOptions? options)
+        private async Task<PingReply> SendPingAsyncInternal<TArg>(
+            TArg getAddressArg,
+            Func<TArg, CancellationToken, ValueTask<IPAddress>> getAddress,
+            int timeout,
+            byte[] buffer,
+            PingOptions? options,
+            CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Previously, cancellations using `SendAsyncCancel()` would get wrapped in `PingException` in the try-catch below. That didn't seem very desirable to me, so I changed it to return a canceled task regardless of how cancellation occurred. Curious what others think.

Other options would be to always wrap, or preserve the wrapping behavior for `SendAsyncCancel()` only.

Reply:
That would probably be a breaking change in the behavior, but since we are implementing "true" cancellation and change the behavior, it might be okay (and we just need to make sure it is documented).

Tagging @stephentoub for more thoughts.

Reply:
> we just need to make sure it is documented

@rzikm Do you mean in the doc comments of the new APIs?

> since we are implementing "true" cancellation and change the behavior, it might be okay

FWIW as a consumer of .NET async APIs, it is frustrating when cancellations get wrapped to look like some other type of exception.

Reply:
> Do you mean in the doc comments of the new APIs?

I meant in release notes that are compiled and list breaking changes. It would be handled by us. AFAIK, we don't list `TaskCancelledException` in docs for async methods.